src/org/daisy/dotify/formatter/impl/segment/MarkerReferenceSegment.java

```diff
@@ -0,0 +1,93 @@
+package org.daisy.dotify.formatter.impl.segment;
+
+import org.daisy.dotify.api.formatter.MarkerReference;
+import org.daisy.dotify.api.formatter.TextProperties;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+/**
+ * Provides a marker reference event object.
+ */
+public class MarkerReferenceSegment implements Segment {
+
+    private final MarkerReference ref;
+    private final TextProperties props;
+    private Function<MarkerReference, String> v = (x) -> "";
+    private String resolved;
+
+    public MarkerReferenceSegment(MarkerReference ref, TextProperties props) {
+        if (ref == null) {
+            throw new IllegalArgumentException();
+        }
+        this.ref = ref;
+        this.props = props;
+    }
+
+    @Override
+    public SegmentType getSegmentType() {
+        return SegmentType.MarkerReference;
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + ref.hashCode();
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null) {
+            return false;
+        }
+        if (getClass() != obj.getClass()) {
+            return false;
+        }
+        MarkerReferenceSegment other = (MarkerReferenceSegment) obj;
+        if (!ref.equals(other.ref)) {
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public String peek() {
+        return resolved == null ? v.apply(ref) : resolved;
+    }
+
+    @Override
+    public String resolve() {
+        if (resolved == null) {
+            resolved = v.apply(ref);
+            if (resolved == null) {
+                resolved = "";
+            }
+        }
+        return resolved;
+    }
+
+    public void setResolver(Function<MarkerReference, String> v) {
+        this.resolved = null;
+        this.v = v;
+    }
+
+    @Override
+    public boolean isStatic() {
+        return false;
+    }
+
+    @Override
+    public Optional<String> getLocale() {
+        return Optional.of(props.getLocale());
```

Comment:
As we discussed in @carl-textalk 's PR the TextProperties locale could be null and that would crash here. Should we continue using Optional.ofNullable here too?

Reply:
Done.